storagemarket/lp2pimpl/net.go

```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
 	}
 
 	if err := cborutil.WriteCborRPC(s, &resp); err != nil {
-		log.Errorf("failed to write ask response: %s", err)
+		log.Warnw("failed to write ask response", "err", err)
 		return
 	}
 }
+
+func (p *DealProvider) handleNewDealStatusStream(s network.Stream) {
+	ctx := context.Background()
+	defer s.Close()
+
+	var q mktnet.DealStatusRequest
+	if err := q.UnmarshalCBOR(s); err != nil {
+		log.Warnw("failed to read DealStatusRequest from incoming stream", "err", err)
+		return
+	}
+
+	dealState, err := p.getMarketsDealState(ctx, q.Proposal)
+	if err != nil {
+		if xerrors.Is(err, storagemarket.ErrDealNotFound) {
+			dealState = &mktssm.ProviderDealState{
+				State:   mktssm.StorageDealUnknown,
+				Message: "deal not found: no deal with proposal cid " + q.Proposal.String(),
+			}
+		} else {
+			log.Errorw("failed to process deal status request", "err", err)
+			dealState = &mktssm.ProviderDealState{
+				State:   mktssm.StorageDealError,
+				Message: "server error: failed to process deal status request",
+			}
+		}
+	}
+
+	signature, err := p.sign(ctx, dealState)
+	if err != nil {
+		log.Errorw("failed to sign deal status response", "err", err)
+		return
+	}
+
+	response := &mktnet.DealStatusResponse{
+		DealState: *dealState,
+		Signature: *signature,
+	}
+
+	if err := cborutil.WriteCborRPC(s, response); err != nil {
+		log.Warnw("failed to write deal status response", "err", err)
+		return
+	}
+}
+
+func (p *DealProvider) getMarketsDealState(ctx context.Context, proposalCid cid.Cid) (*mktssm.ProviderDealState, error) {
+	deal, err := p.prov.DealBySignedProposalCID(ctx, proposalCid)
+	if err != nil {
+		return nil, err
+	}
+
+	signedPropNd, err := cborutil.AsIpld(&deal.ClientDealProposal)
+	if err != nil {
+		return nil, fmt.Errorf("calculating signed deal proposal ipld root: %w", err)
+	}
+	signedPropCid := signedPropNd.Cid()
+
+	dealState := mktssm.ProviderDealState{
+		State:         toLegacyMarketsState(deal),
+		Proposal:      &deal.ClientDealProposal.Proposal,
+		ProposalCid:   &signedPropCid,
+		PublishCid:    deal.PublishCID,
+		DealID:        deal.ChainDealID,
+		FastRetrieval: true,
+	}
+	return &dealState, nil
+}
+
+func (p *DealProvider) sign(ctx context.Context, dealState *mktssm.ProviderDealState) (*crypto.Signature, error) {
```

Comment:
are we doing this only for backward compat ?

Reply:
Yep